opengrok-tools/src/main/python/opengrok_tools/utils/mirror.py

```diff
@@ -0,0 +1,400 @@
+                    format(project_name))
+        return CONTINUE_EXITVAL
+
+    # Check if any of the repositories contains incoming changes.
+    if check_incoming:
+        got_incoming = False
+        for repo in repos:
+            try:
+                if repo.incoming():
+                    logger.debug('Repository {} has incoming changes'.
+                                 format(repo))
+                    got_incoming = True
+                    break
+            except RepositoryException:
+                logger.error('Cannot determine incoming changes for '
+                             'repository {}'.format(repo))
+                return 1
+
+        if not got_incoming:
+            logger.info('No incoming changes for repositories in '
+                        'project {}'.
+                        format(project_name))
+            return CONTINUE_EXITVAL
+
+    if prehook:
+        logger.info("Running pre hook")
+        if run_hook(logger, prehook,
+                    os.path.join(source_root, project_name), proxy,
+                    hook_timeout) != 0:
+            logger.error("pre hook failed for project {}".
+                         format(project_name))
+            return 1
+
+    #
+    # If one of the repositories fails to sync, the whole project sync
+    # is treated as failed, i.e. the program will return 1.
+    #
+    for repo in repos:
+        logger.info("Synchronizing repository {}".
+                    format(repo.path))
+        if repo.sync() != 0:
+            logger.error("failed to synchronize repository {}".
+                         format(repo.path))
+            ret = 1
+
+    if posthook:
+        logger.info("Running post hook")
+        if run_hook(logger, posthook,
+                    os.path.join(source_root, project_name), proxy,
+                    hook_timeout) != 0:
+            logger.error("post hook failed for project {}".
+                         format(project_name))
+            return 1
+
+    return ret
+
+
+def check_project_configuration(multiple_project_config, hookdir=False,
+                                proxy=False):
+    """
+    Check configuration of given project
+    :param multiple_project_config: project configuration dictionary
+    :param hookdir: hook directory
+    :param proxy: proxy setting
+    :return: True if the configuration checks out, False otherwise
+    """
+
+    logger = logging.getLogger(__name__)
+
+    # Quick sanity check.
+    known_project_tunables = [DISABLED_PROPERTY, CMD_TIMEOUT_PROPERTY,
+                              HOOK_TIMEOUT_PROPERTY, PROXY_PROPERTY,
+                              IGNORED_REPOS_PROPERTY, HOOKS_PROPERTY]
+
+    if not multiple_project_config:
+        return True
+
+    for project_name in multiple_project_config.keys():
+        project_config = multiple_project_config.get(project_name)
+        diff = diff_list(project_config.keys(),
+                         known_project_tunables)
```

Comment:
python set supports native difference,
also I could expect different results from diff_list than set difference so the function name is not the best choice here